kong/init.lua

```diff
@@ -147,6 +151,172 @@ local function load_plugins(kong_conf, dao)
+          nil,
+          function(consumer_id)
+            return singletons.dao.consumers:find { id = consumer_id }
+          end,
+          consumer_id)
+
+      else
+        local result = singletons.dao.consumers:find_all { username = consumer_id }
+        if type(result) == "table" and type(result[1]) == "table" then
+          consumer =  result[1]
+        end
+
+        if not consumer then
+          result = singletons.dao.consumers:find_all { custom_id = consumer_id }
+          if type(result) == "table" and type(result[1]) == "table" then
+            consumer = result[1]
+          end
+        end
+      end
+    end
+
+    if consumer then
+      ctx.authenticated_consumer   = consumer
+      ctx.authenticated_credential = {
+        consumer_id = consumer.id
+      }
+
+      ngx.req.set_header(constants.HEADERS.CONSUMER_ID,        consumer.id)
+      ngx.req.set_header(constants.HEADERS.CONSUMER_CUSTOM_ID, consumer.custom_id)
+      ngx.req.set_header(constants.HEADERS.CONSUMER_USERNAME,  consumer.username)
+
+    else
+      return responses.send_HTTP_BAD_REQUEST("consumer not found")
+    end
+  end
+end
+
+
+local function mock_addr(any)
+  local configuration = singletons.configuration
+
+  if any or ngx.var.upstream_scheme == "https" then
+    for _, proxy_server in ipairs(configuration.proxy_servers) do
+      if proxy_server.mock then
+        for _, listener in ipairs(proxy_server.listeners) do
+          if any or listener.ssl then
+            return listener.ip, listener.port
+          end
+        end
+      end
+    end
+  end
+
+  if not any then
+    return mock_addr(true)
+  end
+end
+
+local function mock_data(ctx)
+  local mock = ctx.mock
+
+  local api = ctx.api
+  if api and next(api) then
+    mock.api = api
+
+  else
+    mock.api = null
+  end
+
+  local service = ctx.service
+  if service and next(service) then
+    mock.service = service
+
+  else
+    mock.service = null
+  end
+
+  local route = ctx.route
+  if route and next(route) then
+    mock.route = route
+
+  else
+    mock.route = null
+  end
+
+  local router_matches = ctx.router_matches
+  if router_matches and next(router_matches) then
+    mock.router_matches = router_matches
+
+  else
+    mock.router_matches = null
+  end
+
+  local balancer_address = ctx.balancer_address
+  if balancer_address and next(balancer_address) then
+    mock.balancer_address = balancer_address
+
+  else
+    mock.balancer_address = null
+  end
+
+  local authenticated_consumer = ctx.authenticated_consumer
+  if authenticated_consumer and next(authenticated_consumer) then
+    mock.authenticated_consumer = authenticated_consumer
+
+  else
+    mock.authenticated_consumer = null
+  end
+
+  local authenticated_credential = ctx.authenticated_credential
+  if authenticated_credential and next(authenticated_credential) then
+    mock.authenticated_credential = authenticated_credential
+
+  else
+    mock.authenticated_credential = null
+  end
+end
```

Comment:
This file is becoming a bit too large... Maybe we should move this to `kong.mock` or something?